File: z1_sdk/scripts/example_lowcmd.py
#!/usr/bin/env python3

import time
import rospkg
z1_sdk_path = rospkg.RosPack().get_path('z1_sdk')
import sys
sys.path.append(z1_sdk_path+"/lib")
import unitree_arm_interface
import numpy as np

'''
example_low_cmd.py
An example showing how to control arm motors directly.

Run `roslaunch z1_bringup **_ctrl.launch` first.
'''

# Connect to z1_controller
z1 =  unitree_arm_interface.UnitreeArm("127.0.0.1", 0)
z1.init()


if z1.armState.mode != int(unitree_arm_interface.ArmMode.Passive):
  print("ERROR] Please repower the arm to use lowcmd mode.")
  sys.exit(-1)

# Change to State_lowcmd mode
z1.armCmd.mode = int(unitree_arm_interface.ArmMode.LowCmd)
z1.armCmd.q_d = z1.armState.q
z1.armCmd.gripperCmd.angle = z1.armState.gripperCmd.angle
z1.armCmd.dq_d = np.zeros(6)
z1.armCmd.tau_d = np.zeros(6)

z1.armCmd.Kp = [500, 600, 500, 400, 300, 200]
z1.armCmd.Kd = [5, 5, 5, 5, 5, 5]
z1.sendRecv()

jnt_speed = 0.2

q_d = z1.armState.q_d
for i in range(0, 300):
  q_d[0] += jnt_speed * z1.dt
  z1.armCmd.q_d = q_d
  z1.armCmd.tau_d = np.zeros(6)
  z1.sendRecv()
  time.sleep(z1.dt)

z1.armCmd.mode = int(unitree_arm_interface.ArmMode.Passive)
z1.sendRecv()

File: z1_sdk/examples/example_lowcmd.cpp
#include "unitree_arm_sdk/unitree_arm.h"

/**
 * @example example_lowcmd.cpp
 * An example showing how to control the arm motors directly.
 * 
 * Run `roslaunch z1_bringup **_ctrl.launch` first.
 * @note The gripper is not allowed to contrl in this way.
 */
int main(int argc, char** argv)
{
  /* Connect to z1_controller */
  std::string controller_IP = argc > 1 ? argv[1] : "127.0.0.1";
  UNITREE_ARM_SDK::UnitreeArm z1(controller_IP);
  z1.init();

  if(z1.armState.mode != (mode_t)UNITREE_ARM_SDK::ArmMode::Passive) {
    std::cout << "[ERROR] Please repower the arm to use lowcmd mode.\n";
    exit(-1);
  }

  /* Change to State_Lowcmd mode */
  z1.armCmd.mode = (mode_t)UNITREE_ARM_SDK::ArmMode::LowCmd;
  // Initialize position
  z1.armCmd.q_d = z1.armState.q;
  z1.armCmd.gripperCmd.angle = z1.armState.gripperCmd.angle;
  z1.armCmd.dq_d.fill(0);
  z1.armCmd.tau_d.fill(0);

  // Set control gain
  // The gain of kd & kd has been modified and is now the normal joint factor.
  z1.armCmd.Kp = {500, 600, 500, 400, 300, 200};
  z1.armCmd.Kd = {5, 5, 5, 5, 5, 5};
  z1.sendRecv();

  Timer timer(z1.dt);
  double joint_speed = 0.2;
  for(size_t i(0); i<300; i++)
  {
    z1.armCmd.q_d[0] += joint_speed*z1.dt;
    z1.armCmd.setTau(Vec6::Zero());
    z1.armCmd.gripperCmd.angle -= joint_speed*z1.dt;
    z1.sendRecv();
    timer.sleep();
  }

  /* Set to State_Passive mode */
  z1.armCmd.mode = (mode_t)UNITREE_ARM_SDK::ArmMode::Passive;
  z1.sendRecv();
  return 0;
}